ziti/util/certs.go

```diff
@@ -70,22 +88,25 @@ func IsServerTrusted(host string) (bool, error) {
 	return true, nil
 }
 
-func AreCertsTrusted(host string, certs []byte) (bool, error) {
+func AreCertsTrusted(host string, certs []byte, client http.Client) (bool, error) {
 	tlsConfig := &tls.Config{
 		InsecureSkipVerify: true,
+		RootCAs:            x509.NewCertPool(),
 	}
-
-	if tlsConfig.RootCAs == nil {
-		tlsConfig.RootCAs = x509.NewCertPool()
-	}
-
 	tlsConfig.RootCAs.AppendCertsFromPEM(certs)
 
-	transport := http.DefaultTransport.(*http.Transport).Clone()
-	transport.TLSClientConfig = tlsConfig
-
-	client := http.Client{
-		Transport: transport,
+	if ot, ok := client.Transport.(*http.Transport); ok {
+		it := ot.Clone()
+		it.TLSClientConfig = tlsConfig
+		client.Transport = it
+		defer func() {
+			client.Transport = ot
+		}()
+	} else {
+		client = http.Client{}
+		transport := http.DefaultTransport.(*http.Transport).Clone()
+		transport.TLSClientConfig = tlsConfig
+		client.Transport = transport
 	}
 
 	resp, err := client.Get(fmt.Sprintf("%v/.well-known/est/cacerts", host))
```

Comment:
This function implies that it does 1 thing but does a branching set of things. It checks to see if the supplied HTTP client or the default client if you aren't using a standard *http.Transport and then mutates those clients.

This should be a few separate functions:

- a function to verify a client
- a function to obtain a client
- potentially a function to handle specifically altering the TLS config of the client

It is also odd to take in a client, expecting one, and then return a clone of the (different) default client. I'm not sure what the reasoning is for this. I'm sure there are reasons, but this is triggering code smells.


I do wonder if the http client assembly process is the main issue. Digging into clients we control and built is odd.